chrome/content/zotero/xpcom/itemTreeManager.js

```diff
@@ -0,0 +1,370 @@
+		const allColumns = types.reduce((acc, type) => acc.concat(this._getColumnsByType(type)), []);
+		if (!options) {
+			return allColumns;
+		}
+		if (!Array.isArray(options)) {
+			options = [options];
+		}
+		this._namespacedDataKey(options);
+		/** @type {ItemTreeColumnOptions[]} */
+		const matches = [];
+		for (const opt of options) {
+			const currentMatches = allColumns
+				.filter(col =>
+					// Aoid dataKey collision
+					!matches.map(match => match.dataKey).includes(col.dataKey)
+					// Match all properties
+					&& Object.keys(opt).every(key => col[key] === opt[key])
+				);
+			matches.push(...currentMatches);
+		}
+		return matches;
+	}
+
+	/**
+	 * Check if a column is registered as a custom column
+	 * @param {string} dataKey 
+	 * @returns {boolean} true if the column is registered as a custom column
+	 */
+	isCustomColumn(dataKey) {
+		return !!this._customColumns[dataKey];
+	}
+
+	/**
+	 * A centralized data source for custom columns. This is used by the ItemTreeRow to get data.
+	 * @param {Zotero.Item} item - The item to get data from
+	 * @param {string} dataKey - The dataKey of the column
+	 * @returns {string}
+	 */
+	getCustomCellData(item, dataKey) {
+		const options = this._customColumns[dataKey];
+		if (options && options.dataProvider) {
+			return options.dataProvider(item, dataKey);
+		}
+		return "";
+	}
+
+	/**
+	 * Get columns by type. Only support itemtree and custom for now.
+	 * @param {"itemtree" | "custom"} type 
+	 * @returns {ItemTreeColumnOptions[]}
+	 */
+	_getColumnsByType(type) {
+		type = type || "itemtree";
+		if (type === "itemtree") {
+			return ITEMTREE_COLUMNS.map(opt => Object.assign({}, opt));
+		}
+		if (type === "custom") {
+			return Object.values(this._customColumns).map(opt => Object.assign({}, opt));
+		}
+	}
+
+	/**
+	 * Check if column options is valid.
+	 * If the options is an array, all its children must be valid.
+	 * Otherwise, the validation fails.
+	 * @param {ItemTreeColumnOptions | ItemTreeColumnOptions[]} options 
+	 * @returns {boolean} true if the option(s) are valid
+	 */
+	_validateColumnOption(options) {
+		/**
+		 * Validate column options.
+		 * @param {ItemTreeColumnOptions | ItemTreeColumnOptions[]} options 
+		 * @param {(option: ItemTreeColumnOptions) => boolean} validator - A function that returns true if the option is valid
+		 * @returns {boolean} true if the option(s) are valid
+		 */
+		function validate(options, validator) {
+			if (Array.isArray(options)) {
+				return options.every(opt => validator(opt));
+			}
+			return validator(options);
+		}
+
+		// Check if the input option has duplicate dataKeys
+		const noInputDuplicates = !Array.isArray(options) || !options.find((opt, i, arr) => arr.findIndex(o => o.dataKey === opt.dataKey) !== i);
+		if (!noInputDuplicates) {
+			Zotero.warn(`ItemTree Column options have duplicate dataKey.`);
+		}
+		const requiredProperties = validate(options, (option) => {
+			const valid = option.dataKey && option.label;
```

Comment:
Should require `pluginID`

Reply:
Requiring `pluginID` will prevent ourselves using this API, because we do not have pluginID.

Reply:
We're not going to use this API, but even if we wanted to, we can just pass in a pluginID like `zotero@zotero.org`.

Reply:
Done